// simple-hmac-auth/lib/index.js
"use strict";
//
//  src/index.js
//  Created by Jesse T Youngblood on 11/21/18 at 15:19
//
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.algorithms = exports.sign = exports.canonicalize = exports.AuthError = exports.Server = exports.Client = void 0;
const Client_1 = __importDefault(require("./Client"));
exports.Client = Client_1.default;
const Server_1 = __importDefault(require("./Server"));
exports.Server = Server_1.default;
const AuthError_1 = __importDefault(require("./AuthError"));
exports.AuthError = AuthError_1.default;
const canonicalize_1 = __importDefault(require("./canonicalize"));
exports.canonicalize = canonicalize_1.default;
const sign_1 = require("./sign");
Object.defineProperty(exports, "sign", { enumerable: true, get: function () { return sign_1.sign; } });
Object.defineProperty(exports, "algorithms", { enumerable: true, get: function () { return sign_1.algorithms; } });
module.exports = { Client: Client_1.default, Server: Server_1.default, AuthError: AuthError_1.default, canonicalize: canonicalize_1.default, sign: sign_1.sign, algorithms: sign_1.algorithms };
